hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java

```diff
@@ -299,7 +299,8 @@ protected void completeCompaction(HoodieCommitMetadata metadata, JavaRDD<WriteSt
                                     HoodieTable<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>> table,
                                     String compactionCommitTime) {
     this.context.setJobStatus(this.getClass().getSimpleName(), "Collect compaction write status and commit compaction");
-    List<HoodieWriteStat> writeStats = writeStatuses.map(WriteStatus::getStat).collect();
+    List<HoodieWriteStat> writeStats = metadata.getPartitionToWriteStats().entrySet().stream().flatMap(e ->
+        e.getValue().stream()).collect(Collectors.toList());
```

Comment:
The `metadata` is constructed with the `writeStatuses` already, can you explain what's the difference here ?

Reply:
writeStatuses is defined as JavaRDD<WriteStatus>. And `writeStatuses.map(WriteStatus::getStat).collect()` are RDD computation which are Lazy. Also writeStatuses is persisted before.

When memory cache is invalid, `writeStatuses.map(WriteStatus::getStat).collect()` will do a fully action computation also create new data files with new task ID and Stage ID.

Here use `metadata.getPartitionToWriteStats().xxx` instead of RDD computation to avoid this problem.

Reply:
@zhangyue19921010 
maybe i understand what you mean,
but there may be other changes， thinks that: now i use sparksql to do compaction: writeStatuses.map(WriteStatus::getStat).collect() will call three times,  maybe we also need to do modify for CompactHelpers.createCompactionMetadata.

Reply:
Got it. Trying to get a deeper look.  Thanks :)

Reply:
@zhangyue19921010 good catch! @danny0405 this is a double-writing problem -- de-referencing RDDs second time here (after we've done it in the commit) we creating new files here, which causes divergence.

Reply:
@zhangyue19921010 Great find.  This looks like a real problem.  In that case, do we need to find all such places for getting write stats, to eliminate all similar problems?

Reply:
> @zhangyue19921010 good catch! @danny0405 this is a double-writing problem -- de-referencing RDDs second time here (after we've done it in the commit) we creating new files here, which causes divergence.

@alexeykudinkin  @zhangyue19921010 Thanks for the explanation, i forget that the RDD computation is lazy ~

Reply:
@yihua created a HUDI-3397 to follow-up on this. Will try to take it up this week.